osu.Game/Rulesets/Objects/HitObject.cs

```diff
@@ -215,6 +216,34 @@ public HitSampleInfo CreateHitSampleInfo(string sampleName = HitSampleInfo.HIT_N
 
             return new HitSampleInfo(sampleName);
         }
+
+        public HitObject DeepClone(IDictionary<object, object> referenceLookup)
+        {
+            if (referenceLookup.TryGetValue(this, out object existing))
+                return (HitObject)existing;
+
+            var clone = CreateInstance();
+            referenceLookup[this] = clone;
+
+            clone.CopyFrom(this, referenceLookup);
+
+            return clone;
+        }
+
+        protected virtual HitObject CreateInstance() => new HitObject();
+
+        protected virtual void CopyFrom(HitObject other, IDictionary<object, object> referenceLookup)
+        {
+            StartTime = other.StartTime;
+            Samples = other.Samples.Select(s => s.DeepClone(referenceLookup)).ToList();
+            Kiai = other.Kiai;
+
+            // HitWindows is immutable, so we can copy by reference
+            HitWindows = other.HitWindows;
+
+            nestedHitObjects.Clear();
+            nestedHitObjects.AddRange(other.nestedHitObjects.Select(h => h.DeepClone(referenceLookup)));
+        }
```

Comment:
I dunno. This is to my eyes nothing short of shit.

We can't expect every ruleset to implement this stuff, let alone get it right.

Case in point: test failures. from missing `CreateInstance` implementations or incorrect `CopyFrom` implementations.

Reply:
agreed. we can try to find simplifications or move to automapper, but the end result will always be that there's annoying complexity around cloning objects. simplifying hitobjects and related objects themselves internally would go a long way in allowing much more straightforward cloning logic. maybe go in that direction before trying to implement it?